### PR TITLE
Core/Spells: Implement SPELL_ATTR7_INTERRUPT_ONLY_NONPLAYER

### LINKED ISSUE #14972: [3.3.5] Missing interrupt spellcasting on Hammer Of Justice, Maim and Bash
as per title.
In patch 3.x, those 3 spells have a linked interrupt spellcasting effect, which is actually used vs creatures ( dungeon/raid bosses), commonly when stun effect of the spell cannot land cause of immunity.

This mechanic was changed in patch 3.2.0 so that the interrupt effect was applied only to mobs, not to players.

below the patch to make it work. I will push it, just if anyone want to test it before i do tomorrow afternoon when i'll have again a 3.3.5 client :)

```cpp
diff --git a/src/server/scripts/Spells/spell_generic.cpp b/src/server/scripts/Spells/spell_generic.cpp
index 68c115f..eea7bdd 100644
--- a/src/server/scripts/Spells/spell_generic.cpp
+++ b/src/server/scripts/Spells/spell_generic.cpp
@@ -4079,6 +4079,47 @@ public:
     }
 };
 
+// PVE additional interrupt effect for Maim (-22570), Bash (-5211), Hammer of Justice (-853)
+enum GenericInterruptSpells
+{
+    SPELL_INTERRUPT = 32747
+};
+
+class spell_gen_interrupt : public SpellScriptLoader
+{
+public:
+    spell_gen_interrupt() : SpellScriptLoader("spell_gen_interrupt") { }
+
+    class spell_gen_interrupt_SpellScript : public SpellScript
+    {
+        PrepareSpellScript(spell_gen_interrupt_SpellScript);
+
+        bool Validate(SpellInfo const* /*spellInfo*/) override
+        {
+            if (!sSpellMgr->GetSpellInfo(SPELL_INTERRUPT))
+                return false;
+            return true;
+        }
+
+        void HandleScript()
+        {
+            if (Unit* target = GetHitCreature())
+                    GetCaster()->CastSpell(target, SPELL_INTERRUPT, true);
+        }
+
+        void Register() override
+        {
+            BeforeHit += SpellHitFn(spell_gen_interrupt_SpellScript::HandleScript);
+        }
+    };
+
+    SpellScript* GetSpellScript() const override
+    {
+        return new spell_gen_interrupt_SpellScript();
+    }
+};
+
 void AddSC_generic_spell_scripts()
 {
     new spell_gen_absorb0_hitlimit1();
@@ -4115,6 +4156,7 @@ void AddSC_generic_spell_scripts()
     new spell_gen_gadgetzan_transporter_backfire();
     new spell_gen_gift_of_naaru();
     new spell_gen_gnomish_transporter();
+    new spell_gen_interrupt();
     new spell_gen_lifeblood();
     new spell_gen_lifebloom("spell_hexlord_lifebloom", SPELL_HEXLORD_MALACRASS_LIFEBLOOM_FINAL_HEAL);
     new spell_gen_lifebloom("spell_tur_ragepaw_lifebloom", SPELL_TUR_RAGEPAW_LIFEBLOOM_FINAL_HEAL);
```

and the sql part:
```sql
DELETE FROM `spell_script_names` WHERE `scriptname` = 'spell_gen_interrupt';
INSERT INTO `spell_script_names` VALUES 
(-22570,'spell_gen_interrupt'), -- Maim
(-5211,'spell_gen_interrupt'),  -- Bash
(-853,'spell_gen_interrupt');   -- Hammer of Justice
```

### PR DESCRIPTION
**Pull request by mik1893**
**Changes proposed**:
- Implement non-player interrupt mechanic for spells like Hammer of Justice, Maim and Bash

**Target branch(es)**: 335

**Issues addressed**: Fixes #14972 

**Tests performed**: (Does it build, tested in-game, etc) - Build - tested

**Known issues and TODO list** - none

### CODE REVIEW

**Comment by mik1893** [comment]
let me check and will fix that

File: src/server/game/Spells/Spell.h [hunk 1]
```diff
@@ -36,6 +36,8 @@ class ByteBuffer;
 
 #define SPELL_CHANNEL_UPDATE_INTERVAL (1 * IN_MILLISECONDS)
 
+#define SPELL_INTERRUPT_NONPLAYER 32747
```
Comment (Treeston) [inline src/server/game/Spells/Spell.h]: `#define` is bad and you should feel bad.

`enum` or `static const uint32` please.

File: src/server/game/Spells/Spell.cpp [hunk 2]
```diff
@@ -6857,6 +6857,11 @@ void Spell::DoAllEffectOnLaunchTarget(TargetInfo& targetInfo, float* multiplier)
     if (!unit)
         return;
 
+    // Check for SPELL_ATTR7_INTERRUPT_ONLY_NONPLAYER and eventually cast the interrupt spell if the target is a non-player
```
Comment (Shauren) [inline src/server/game/Spells/Spell.cpp]: I disagree with where this code is placed - Silencing Shot has travel time and should interrupt on impact, not when started

Comment (mik1893) [inline src/server/game/Spells/Spell.cpp]: ok - on spell hit then

File: src/server/game/Spells/Spell.cpp [hunk 3]
```diff
@@ -2676,6 +2677,11 @@ SpellMissInfo Spell::DoSpellHitOnUnit(Unit* unit, uint32 effectMask, bool scaleA
         if (effectMask & (1 << effectNumber))
             HandleEffects(unit, NULL, NULL, effectNumber, SPELL_EFFECT_HANDLE_HIT_TARGET);
 
+    // Check for SPELL_ATTR7_INTERRUPT_ONLY_NONPLAYER and eventually cast the interrupt spell if the target is a non-player
+    if (m_spellInfo->HasAttribute(SPELL_ATTR7_INTERRUPT_ONLY_NONPLAYER))
+        if (unit->GetTypeId() != TYPEID_PLAYER)
```
Comment (GitVerge) [inline src/server/game/Spells/Spell.cpp]: why not merge the 2 checks in a single if?

Comment (tkrokli) [inline src/server/game/Spells/Spell.cpp]: The point is to skip the second check if the first check returns `false`.

Comment (Treeston) [inline src/server/game/Spells/Spell.cpp]: That is how C++ already works. Please don't comment if you have no idea what you are talking about.

Comment (Treeston) [inline src/server/game/Spells/Spell.cpp]: I agree the checks should be merged btw.

```cpp
if (m_spellInfo->HasAttribute(SPELL_ATTR7_INTERRUPT_ONLY_NONPLAYER) && unit->GetTypeId() != TYPEID_PLAYER)
```

File: src/server/game/Spells/Spell.h [hunk 4]
```diff
@@ -218,6 +218,8 @@ enum SpellEffectHandleMode
 
 typedef std::list<std::pair<uint32, ObjectGuid>> DispelList;
 
+static const uint32 SPELL_INTERRUPT_NONPLAYER = 32747;
```
Comment (Treeston) [inline src/server/game/Spells/Spell.h]: Not sure if this really belongs in the header. Seems random.

File: src/server/game/Spells/Spell.cpp (hunk 3 above)
Comment (Keader) [inline src/server/game/Spells/Spell.cpp]: yeap, has no reason for dont merge :)

**Comment by mik1893** [comment]
i'm changing it soon - will merge the check - need to understand why it doesn't cast the spell in that function though

File: src/server/game/Spells/Spell.h (hunk 4 above)
Comment (mik1893) [inline src/server/game/Spells/Spell.h]: the number is just in description, it s not in dbc or anywhere else... what should i do with that?

**Comment by mik1893** [comment]
ok, i think it logically should stay in the triggers execution.

@Keader can you retest this?

**Comment by Rushor** [comment]
will do as fast as possible, if not anyone else did it faster  - thank you

**Comment by Keader** [comment]
Interrupt effect should be applied even if Hammer, Silence, Bash, etc get immune?
Like Lady Deathwhisper (Icecrown Citadel). Lady is immune to stun/silence, but is not immune to interrupt effects.
Spells should work on her? if should, pr is not working ^^

**Comment by Rushor** [comment]
> Like Lady Deathwhisper (Icecrown Citadel). Lady is immune to stun/silence, but is not immune to interrupt effects.

yeah that's a perfect encounter to test the issue - same here sadly it's not working :(

**Comment by mik1893** [comment]
ok... i think i understood why - i will update asap

**Comment by Rushor** [comment]
thank you very much !

**Comment by mik1893** [comment]
i didnt forget about this - just dont have time to properly test it, will do it in the weekend most likely

**Comment by mik1893** [comment]
FIXED FIXED FIXED 
thanks @Shauren even if u dont deserve it

File: src/server/game/Spells/Spell.cpp [hunk 5]
```diff
@@ -2287,6 +2287,10 @@ void Spell::DoAllEffectOnTarget(TargetInfo* target)
     if (!caster)
         return;
 
+    // Check for SPELL_ATTR7_INTERRUPT_ONLY_NONPLAYER before checking the miss conditions
+    if (m_spellInfo->HasAttribute(SPELL_ATTR7_INTERRUPT_ONLY_NONPLAYER) && unit->GetTypeId() != TYPEID_PLAYER)
```
Comment (Shauren) [inline src/server/game/Spells/Spell.cpp]: just a note, you probably should place if AFTER handling effects - the blizzlike effect is that if the stun interrupts the cast (if not immune), it should happen before this spell and not show in combat log as being interrupted by 32747

Comment (Treeston) [inline src/server/game/Spells/Spell.cpp]: Shouldn't there just be a check on "is the spell being immuned"? If yes, cast.

Comment (mik1893) [inline src/server/game/Spells/Spell.cpp]: technically it shouldnt check for it to be immuned 
Stuns the target for $d and interrupts non-player spellcasting for $32747d.
So yeah shauren is likely right. Let me move it after the handling the effect.

**Pull request merged**